tools/serve/serve.py

```diff
@@ -204,6 +207,65 @@ def _script_replacement(self, key, value):
         return None
 
 
+class HtmlScriptInjectorHandlerWrapper:
+    def __init__(self, inject="", wrap=None):
+        self.inject = inject
+        self.wrap = wrap
+
+    def __call__(self, request, response):
+        self.wrap(request, response)
+        # If the response content type isn't html, don't modify it.
+        supported_mime_types = {b"text/html", b"application/xhtml+xml"}
+        if not isinstance(response.headers, ResponseHeaders) or not response.headers.get("Content-Type")[0] in supported_mime_types:
+            return response
+
+        # Skip injection on custom streaming responses.
+        if not isinstance(response.content, (bytes, str, IOBase)) and not hasattr(response, "read"):
+            return response
+
+        # Otherwise, inject the script after the document's doctype.
+        data = "".join(item.decode(response.encoding) for item in response.iter_content(read_file=True))
+
+        # Tokenize and find the position of the first content (e.g. after the
+        # doctype if one is present).
+        token_types = html5parser.tokenTypes
+        after_tags = {"html", "head"}
+        before_tokens = {token_types["EndTag"], token_types["EmptyTag"],
+                         token_types["Characters"]}
+        error_tokens = {token_types["ParseError"]}
+
+        tokenizer = html5parser._tokenizer.HTMLTokenizer(data)
+        stream = tokenizer.stream
+        offset = 0
+        error = False
+        for item in tokenizer:
```

Comment:
This loop here is where all of the good stuff happens. Is it possible to test this with some different input to show that it inserts in the right place? Unfortunately I don't see any tests for any *Handler, so I don't know what to suggest. @jgraham thoughts?

Reply:
I think we could move this part of the code to a standalone function which could be called with simple test cases. I'll put together a proof of concept.

Reply:
I've done this and added tests of the main different injection scenarios. Let me know if you think there's anything missing.